### PR TITLE
[CPDLP-2107] Add GET /api/v3/participants/ecf endpoint

### PR DESCRIPTION
**Pull request by leandroalemao**
### Context

We would like to add the participant’s ECF endpoint to v3.

- Ticket: [CPDLP-2107](https://dfedigital.atlassian.net/browse/CPDLP-2107)

### Changes proposed in this pull request

Add GET /api/v3/participants/ecf endpoint.

[CPDLP-2107]: https://dfedigital.atlassian.net/browse/CPDLP-2107?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/services/api/v3/ecf/participants_query.rb [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+                       partnerships: {
+                         lead_provider_id: lead_provider.id,
+                         challenged_at: nil,
+                         challenge_reason: nil,
+                         pending: false,
+                       },
+                     },
+                   )
+
+          scope = User
```
Comment (cwrw) [inline app/services/api/v3/ecf/participants_query.rb:28]: could we add the selects back please as they reduced the query times? (can reuse a lot from v1)?

File: docs/source/release-notes.html.md [hunk 2]
```diff
@@ -37,6 +37,17 @@ Note, providers may need to update IDs held on their systems. The DfE will conta
 
 While the expected volume of future instances is expected to be very low (where single participants are registered with multiple `participant_ids`), providers who identify issues should contact the DfE via existing support channels. The DfE will fix errors by 'retiring' one of the IDs, then associating all records and data under the remaining ID. The DfE will inform providers as to which `participant_id` they will need to record on their systems.
 
+## 14th March 2023
```
Comment (cwrw) [inline docs/source/release-notes.html.md:40]: ah thanks for adding back!

File: app/serializers/api/v3/ecf/participant_serializer.rb [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        set_type :participant
+
+        attribute :full_name
+
+        attribute(:teacher_reference_number) do |object|
+          object.teacher_profile&.trn
+        end
+
+        attribute(:updated_at) do |object|
+          object.updated_at.rfc3339
```
Comment (cwrw) [inline app/serializers/api/v3/ecf/participant_serializer.rb:80]: I wonder if we should keep the same v1 list of updated_at and pick the max?  🤔 
```
[
  induction_record.participant_profile_updated_at,
  induction_record.user_updated_at,
  induction_record.participant_identity_updated_at,
  induction_record.updated_at,
].compact.max.rfc3339
```

**Review by cwrw — COMMENTED** [review]
amazing effort, a few comments but looking great, thanks Leandro 👍

File: app/serializers/api/v3/ecf/participant_serializer.rb [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        attribute(:ecf_enrolments) do |object, params|
+          object.participant_profiles.select { |pp| [ParticipantProfile::ECT.name, ParticipantProfile::Mentor.name].include?(pp.type) }.map { |profile|
+            latest_induction_record = profile.induction_records.first
+
+            next unless params[:cpd_lead_provider] && latest_induction_record.present? && latest_induction_record.induction_programme&.partnership&.lead_provider&.cpd_lead_provider == params[:cpd_lead_provider]
+
+            {
+              training_record_id: profile.id,
+              email: latest_induction_record.preferred_identity&.email.presence || object.email,
+              mentor_id: latest_induction_record.participant_profile.ect? ? latest_induction_record.mentor_profile&.participant_identity&.user_id : nil,
```
Comment (cwrw) [inline app/serializers/api/v3/ecf/participant_serializer.rb:92]: ah I think I spotted some n+1s here will need to address as well (in later pages :))

**Review by cwrw — COMMENTED** [review]
Looking great, well done Leandro! I have 1 suggestion and another comment on the action endpoints, otherwise all good

File: app/services/api/v3/ecf/participants_query.rb [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+                    .select(*necessary_fields)
```
[the remark this answers is not in the document]
Comment (leandroalemao) [reply app/services/api/v3/ecf/participants_query.rb]: I removed the select completely as the results are getting returned faster without it, at least in my machine.

File: app/serializers/api/v3/ecf/participant_serializer.rb [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+              school_urn: latest_induction_record.induction_programme&.school_cohort&.school&.urn,
+              participant_type: profile.ect? ? :ect : :mentor,
+              cohort: latest_induction_record.induction_programme&.school_cohort&.cohort&.start_year&.to_s,
+              training_status: latest_induction_record&.training_status,
+              participant_status: participant_status(induction_record: latest_induction_record),
+              teacher_reference_number_validated: trn(profile).nil? ? nil : validated_trn(profile).present?,
+              eligible_for_funding: eligible_for_funding?(profile),
+              pupil_premium_uplift: profile.pupil_premium_uplift,
+              sparsity_uplift: profile.sparsity_uplift,
+              schedule_identifier: profile.schedule&.schedule_identifier,
```
Comment (cwrw) [inline app/serializers/api/v3/ecf/participant_serializer.rb:112]:
```suggestion
              schedule_identifier: latest_induction_record.schedule&.schedule_identifier,
```

**Review by cwrw — APPROVED** [review]
🙌 nice one, just one last comment

File: app/services/api/v3/ecf/participants_query.rb [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+                         challenged_at: nil,
+                         challenge_reason: nil,
+                       },
+                     },
+                   )
+
+          scope = User
+                    .includes(:participant_identities, :teacher_profile, participant_profiles: [:participant_profile_states, :schedule, :teacher_profile, :ecf_participant_eligibility, :ecf_participant_validation_data, { induction_records: [:preferred_identity, :schedule, :delivery_partner, { induction_programme: { partnership: [lead_provider: :cpd_lead_provider], school_cohort: %i[school cohort] } }, { mentor_profile: :participant_identity }] }])
+                    .eager_load(participant_profiles: [:induction_records])
+                    .joins(left_outer_join_preferred_identities)
```
Comment (cwrw) [inline app/services/api/v3/ecf/participants_query.rb:30]: @leandroalemao to check if we need the left outer joins (might do)

File: app/controllers/api/v3/ecf/participants_controller.rb [hunk 8]
```diff
@@ -53,7 +53,7 @@ def ecf_participant_query
         def render_from_service(service, serializer, params: {})
           if service.valid?
             induction_record = service.call
-            render json: serializer.new(induction_record.user, params:).serializable_hash
+            render json: serializer.new(induction_record.user, params: params.merge(induction_record:)).serializable_hash
```
Comment (cwrw) [inline app/controllers/api/v3/ecf/participants_controller.rb:56]: I don't think we need to pass this `induction_record` param in anymore

**Review by cwrw — APPROVED** [review]

File: app/serializers/api/v3/ecf/participant_serializer.rb [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+                  reason: latest_participant_profile_state.reason,
+                  date: latest_participant_profile_state.created_at.rfc3339,
+                }
+              end
+            end
+          end
+
+          def deferral(profile:, cpd_lead_provider:, latest_induction_record:)
+            if latest_induction_record.training_status_deferred?
+              latest_participant_profile_state = profile.participant_profile_states.sort_by(&:created_at).find { |pps| pps.state == ParticipantProfileState.states[:deferred] && pps.cpd_lead_provider_id == cpd_lead_provider.id }
```
Comment (cwrw) [inline app/serializers/api/v3/ecf/participant_serializer.rb:48]:
```suggestion
              latest_participant_profile_state = profile.participant_profile_states.sort_by(&:created_at).reverse!.find { |pps| pps.state == ParticipantProfileState.states[:deferred] && pps.cpd_lead_provider_id == cpd_lead_provider.id }
```

**Comment by cwrw** [comment]
@leandroalemao could we amend the request specs for actions to check the response that deferrals/withdrawals are set, as I think it's only checking minor things form v1?

File: app/services/api/v3/ecf/participants_query.rb [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+        def left_outer_join_participant_profiles
+          "LEFT OUTER JOIN participant_profiles ON participant_profiles.id = induction_records.participant_profile_id"
+        end
+
+        def left_outer_join_participant_identities
+          "LEFT OUTER JOIN participant_identities ON participant_identities.id = participant_profiles.participant_identity_id"
+        end
+
+        def left_outer_join_participant_profile_states
```
Comment (cwrw) [inline app/services/api/v3/ecf/participants_query.rb:91]: chatted to @leandroalemao: we need to select all profile states regardless of latest

**Review by cwrw — APPROVED** [review]
🙌

**Pull request merged**